src/containers/rbtree/rbtree.h

```diff
@@ -0,0 +1,794 @@
+
+  iterator begin() noexcept;
+  iterator end() noexcept;
+
+  std::pair<iterator, bool> insert(const value_type& value) noexcept;
+  std::pair<iterator, bool> insert(const K& key, const T& obj);
+  std::pair<iterator, bool> insert_or_assign(const K& key, const T& obj);
+  void erase(iterator pos);
+  void erase(const value_type& value);
+  void merge(rbtree& other);
+  T& at(const K& key);
+  T& operator[](const K& key);
+  bool contains(const K& key);
+  size_type size();
+  void print_tree() { print_tree(m_root, 0); }
+  void clear();
+
+  node* minimum(node* n);
+  node* find_node(const K& key);
+
+ private:
+  void rotate_left(node* pos);
+  void rotate_right(node* pos);
+  void insert(const value_type& value, iterator& iter, bool& inserted);
+  void insert_node(node* temp, iterator& iter, bool& inserted);
+  void clear(node* node);
+  node* copy_tree(node* node);
+  void transfer_node(node* deleted_node, node* inserted_node) noexcept;
+  bool is_leaf(node* n);
+  void set_new_grandfather(node* parent, node* uncle, node* grandfather,
+                           node*& pos);
+  node* maximum(node* n);
+  node* sibling(node* n);
+  void fix_insertion(node* pos);
+  void replace_node(node* n, node* child) noexcept;
+  void delete_case1(node* n);
+  void delete_case2(node* n);
+  void delete_case3(node* n);
+  void delete_case4(node* n);
+  void delete_case5(node* n);
+  void delete_case6(node* n);
+
+  void print_tree(node* current, int depth) {
+    if (current == nullptr) {
+      return;
+    }
+    print_tree(current->right_, depth + 1);
+    std::cout << std::setw(depth * 4) << "";
+    std::cout << (current->mColor == 'R' ? "R" : "B") << current->value_.first
+              << std::endl;
+    print_tree(current->left_, depth + 1);
+  }
+
+ private:
+  node* m_root;
+  size_type m_size;
+};
+
+template <typename K, typename T>
+inline rbtree<K, T>::rbtree_iterator::rbtree_iterator() noexcept
+    : current_(nullptr) {}
+
+template <typename K, typename T>
+inline rbtree<K, T>::rbtree_iterator::rbtree_iterator(node* node) noexcept
+    : current_(node) {}
+
+template <typename K, typename T>
+inline typename rbtree<K, T>::reference
+rbtree<K, T>::rbtree_iterator::operator*() const {
+  return current_->value_;
+}
+
+template <typename K, typename T>
+inline typename rbtree<K, T>::pointer
+rbtree<K, T>::rbtree_iterator::operator->() const {
+  return &(current_->value_);
+}
+
+template <typename K, typename T>
+inline typename rbtree<K, T>::rbtree_iterator&
+rbtree<K, T>::rbtree_iterator::operator++() {
+  if (current_ != nullptr) {
+    if (current_->right_) {
+      current_ = current_->right_;
+      while (current_->left_) {
+        current_ = current_->left_;
+      }
+    } else {
+      node* temp = current_->parent_;
+      while (temp != nullptr && current_ == temp->right_) {
+        current_ = temp;
+        temp = temp->parent_;
+      }
+      current_ = temp;
+    }
+  }
+  return *this;
+}
```

Comment:
1. Лучше не использовать такие вложенности. Это не "чистый Си" и у нас здесь нет требования - "один `return` на функцию". Лучше проверять значения и если они не подходят, то делать возврат. Например, эту функцию можно переписать вот так и она сразу станет чуточку проще и понятнее.

```suggestion
template <typename K, typename T>
inline typename rbtree<K, T>::rbtree_iterator&
rbtree<K, T>::rbtree_iterator::operator++() {
  if (current_ == nullptr) return;
  
  if (current_->right_) {
      current_ = current_->right_;
      while (current_->left_) {
        current_ = current_->left_;
      }
    } else {
      node* temp = current_->parent_;
      while (temp != nullptr && current_ == temp->right_) {
        current_ = temp;
        temp = temp->parent_;
      }
      current_ = temp;
    }

  return *this;
}
```